src/ut_control_plane.c

```diff
@@ -0,0 +1,74 @@
+#include <libwebsockets.h>
+#include <stdio.h>
+#include <ut_control_plane.h>
+
+char *test_message="HI FROM SERVER";
+
+static int callback_echo(struct lws *wsi, enum lws_callback_reasons reason, void *user, void *in, size_t len)
+{
+    switch (reason)
+    {
+    case LWS_CALLBACK_ESTABLISHED:
+        printf("Client connected\n");
+        break;
+
+    case LWS_CALLBACK_RECEIVE:
+        printf("LWS_CALLBACK_RECEIVE\n");
+        // Echo back received message
+        printf("Received message: %.*s\n", (int)len, (char *)in);
+        lws_write(wsi, in, len, LWS_WRITE_TEXT);
+        break;
+
+    default:
+        break;
+    }
+
+    return 0;
+}
+
+static struct lws_protocols protocols[] = {
+    {
+        "echo-protocol",    // protocol name
+        callback_echo,      // callback function
+        0,                  // per-session data size
+        0,                  // rx buffer size
+        0,                  // id
+        NULL,               // reserved
+        0                   // tx packet size
+    },
+    { NULL, NULL, 0, 0 }    // end of list
+};
+
+void ut_control_panel()
+{
+    struct lws_context_creation_info info;
+    struct lws_context *context;
+    const char *address = "localhost";
+    int port = 8080;
+
+    memset(&info, 0, sizeof(info));
+    info.port = port;
+    info.iface = NULL;
+    info.protocols = protocols;
+
+    context = lws_create_context(&info);
+    if (!context)
+    {
+        fprintf(stderr, "Error creating libwebsockets context\n");
+        return;
+    }
+
+    printf("WebSocket server started at %s:%d\n", address, port);
+
+    while (1)
```

Comment:
this can't wait forever, it must look for messages.

- message get's sent to the thread
- Thread is waiting on any message, and timeout
- timeout occurs, should I exit? if so then exit otherwise loop.
- thread wakes, start to process the message.
- looks at the callback list and the key value pair in the message against the callback list,.
- calls the callback required if patch
- no match, free memory, loop to wait on timeout.

- Define a structure for the message 

```c
typedef struct
{
 eMessage_t status;
 char *message;
 uint32_t size;
}myMessage_t;
```

```c

## Message to quite the thread, in the deinit function()
myMessage_t newMessage;
newMessage.status = EXIT_REQUESTED;
sendMessageToThread( &newMessage );

blockWaitForThreadToExit(); # Block waiting for the thread to exit
..
exit function()

```

```c
myMessage_t newMessage;
newMessage.status = DATA_RECIEVED;
newMessage.data = myblockofmemory;
newMessage.size = myblocksize;
sendMessageToThread( &newMessage );
```

Reply:
using kvp, look up `list[index].key` in message, if found send message to module and continue loop

end of loop
free memory from message.

Reply:
on exiting the thread, set a volatile, semaphore flag or mutex to tell the sender no more messages allowed, and just quit with error if anyone sends more messages.

- loop round the rest of the message Q, ensure that all memory is freed from every message, but don't call callbacks.